Add Remote calendar integration

Thomas55555 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a remote calendar integration. It allows to add a read-only calendar in ics format. Like suggested [here](https://github.com/home-assistant/core/pull/126862#issuecomment-2384812485).
It's basically a copy of the local calendar and the write parts are removed. I also added a `DataUpdateCoordinator` and removed the local storage, so that the data comes from the source directly.
<img width="431" alt="image" src="https://github.com/user-attachments/assets/a6869561-173f-46ba-be2f-4e46a16cd2e1" />
<img width="1722" alt="image" src="https://github.com/user-attachments/assets/61b3d510-3f9b-453f-994c-05bf9f8d5f99" />



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/37568
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Thomas55555 committed:
Add remote_calendar with storage

Thomas55555 committed:
Use coordinator and remove storage

Thomas55555 committed:
cleanup

Thomas55555 committed:
cleanup

Thomas55555 committed:
Merge branch 'home-assistant:dev' into remote_calendar

Thomas55555 committed:
remove init from config_flow

Thomas55555 committed:
add some tests

Thomas55555 committed:
some fixes

Thomas55555 committed:
test-before-setup

Thomas55555 committed:
fix error handling

Thomas55555 committed:
remove unneeded code

Thomas55555 committed:
fix updates

Thomas55555 committed:
load calendar in the event loop

Thomas55555 committed:
allow redirects

Thomas55555 committed:
test_update_failed

Thomas55555 committed:
tests

allenporter commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+        self._abort_if_unique_id_configured()
+        client = get_async_client(self.hass)
+        _LOGGER.debug("User input in fetch url: %s", user_input)
+        if user_input is not None:
```
I believe this should always be true here given the check on line 34

allenporter commented on homeassistant/components/remote_calendar/coordinator.py:
```diff
@@ -0,0 +1,74 @@
+        )
+        self._etag = None
+        self._client = get_async_client(hass)
+        self._url = config_entry.data["url"]
```
Use CONF_URL here

allenporter reviewed (changes requested):
Awesome contribution! This is requested quite a bit so I know folks will definitely like this. 👍🏼  Very good direction, and i just took an initial pass to get the review started (didn't review the tests in detail yet)

Thomas55555 commented on homeassistant/components/remote_calendar/calendar.py:
```diff
@@ -0,0 +1,106 @@
+        self._event: CalendarEvent | None = None
+        self._attr_name = name
+        self._attr_unique_id = unique_id
+        self._attr_should_poll = True
```
I had problems with polling not working. So I added this, and forgot to remove it.

Thomas55555 commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {vol.Required(CONF_CALENDAR_NAME): str, vol.Required(CONF_URL): str}
```
ruff is changing it back all the time

allenporter replied:
I think adding a `,` after the url field will fix that.

Thomas55555 commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+        client = get_async_client(self.hass)
+        _LOGGER.debug("User input in fetch url: %s", user_input)
+        if user_input is not None:
+            headers: dict = {}
```
yes

Thomas55555 committed:
address review

Thomas55555 commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+                        IcsCalendarStream.calendar_from_ics, res.text
+                    )
+                except CalendarParseError as err:
+                    errors["base"] = "no_calendar_found"
```
Yes, would it be okay, to import
`"[%key:component::local_calendar::config::error::invalid_ics_file%]"`
Or is this a dependency we don't want?

Thomas55555 committed:
use error from local_calendar

Thomas55555 committed:
adress more comments

Thomas55555 committed:
remove unique_id

Thomas55555 commented:
> Awesome contribution! This is requested quite a bit so I know folks will definitely like this. 👍🏼 Very good direction, and i just took an initial pass to get the review started (didn't review the tests in detail yet)

Thanks. Tests aren't complete anyway. I just started, with the basics. During the review there are a lot of changes anyway. So it's not really worth to test everything in the beginning.

Thomas55555 commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+            )
+        errors: dict = {}
+        _LOGGER.debug("User input: %s", user_input)
+        await self.async_set_unique_id(slugify(user_input[CONF_URL]))
```
ah okay, we need at least a unique enitiy_id

Thomas55555 replied:
For the config entry:
maybe we can hash the ics file and use this as unique id?

Thomas55555 committed:
add unique entity_id

Thomas55555 committed:
add excemption

Thomas55555 committed:
abort_entries_match

frenck commented:
I've marked this PR as draft, as the CI is failing.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

allenporter commented on homeassistant/components/remote_calendar/calendar.py:
```diff
@@ -0,0 +1,96 @@
+        """Initialize RemoteCalendarEntity."""
+        super().__init__(coordinator)
+        self._attr_name = name
+        self._attr_unique_id = name
```
My impression from https://developers.home-assistant.io/docs/entity_registry_index#unique-id-requirements is this is not a valid unique id.

allenporter commented on homeassistant/components/remote_calendar/calendar.py:
```diff
@@ -0,0 +1,96 @@
+    """Return a CalendarEvent from an API event."""
+    start: datetime | date
+    end: datetime | date
+    if isinstance(event.start, datetime) and isinstance(event.end, datetime):
```
This logic for coercing dates in local calendar is here to cleanup a bug for how old calendar events were created https://github.com/home-assistant/core/pull/90970 Is there a specific reason this is all here or just copy/paste? If the latter, then remove and only include what is needed to comply with the `CalendarEvent` (e.g. ensure datetimes have a timezone)

allenporter commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -0,0 +1,76 @@
+            errors["base"] = "unknown_url_type"
+            _LOGGER.debug("ValueError: %s", err)
+        else:
+            res.raise_for_status()
```
What kind of errors does this raise? (presumably they need to be caught somewhere). Is this testsd?

Thomas55555 replied:
I can't simulate an error here. When the destination is unreachable or the client has no internet, this is catched already before. So no need to keep it (was also copied and pasted).

Thomas55555 replied:
I did some research. Yes, it has to go in the try block. There is a library called respx, which does the same for httpx like aioresponses. I've changed it (not submitted, yet).

allenporter commented on homeassistant/components/remote_calendar/coordinator.py:
```diff
@@ -0,0 +1,72 @@
+                    translation_placeholders={"err": str(err)},
+                ) from err
+            else:
+                return await self.hass.async_add_executor_job(
```
This appears to be called twice. Can this return from the previous call?

Thomas55555 replied:
https://github.com/home-assistant/core/pull/138862/commits/adaacdb0209b2e6ecd025a48ded57e58ddbfee1f

allenporter commented on homeassistant/components/remote_calendar/coordinator.py:
```diff
@@ -0,0 +1,72 @@
+                translation_key="unable_to_fetch",
+                translation_placeholders={"err": str(err)},
+            ) from err
+        else:
```
The else nesting is not needed since all exception blocks are returning/raising. Remove the else and de-dent the code blocks to save horizontal whitespace and have less nesting here and elsewhere. 

(The reason it is there in the config flow is because the except cases re-render the form)

Thomas55555 replied:
https://github.com/home-assistant/core/pull/138862/commits/4969c0cf72f7a9a9bab2c1a6876327716898c448

allenporter reviewed (changes requested):
I think its reasonable to work on the tests at this point. (not that much will change for the rest of the review)

Thomas55555 committed:
unique_id

Thomas55555 committed:
add ,

Thomas55555 committed:
cleanup

Thomas55555 committed:
deduplicate call

Thomas55555 committed:
don't raise for status end de-nest

allenporter commented on homeassistant/components/remote_calendar/calendar.py:
```diff
@@ -42,12 +41,12 @@ class RemoteCalendarEntity(
-        self._attr_name = name
-        self._attr_unique_id = name
+        self._attr_name = entry.data[CONF_CALENDAR_NAME]
+        self._attr_unique_id = entry.entry_id
```
Ah misread, thank you.

Thomas55555 committed:
multiline

Thomas55555 committed:
test

Thomas55555 committed:
tests

Thomas55555 commented on homeassistant/components/remote_calendar/calendar.py:
```diff
@@ -71,23 +70,15 @@ async def async_get_events(
         summary=event.summary,
-        start=start,
-        end=end,
+        start=dt_util.as_local(event.start)
```
https://github.com/home-assistant/core/pull/138862/commits/e3908e5e94a2403f24f82ba9121687e5f03bc98f

Thomas55555 committed:
use raise_for_status again

allenporter commented on tests/components/remote_calendar/test_calendar.py:
```diff
@@ -0,0 +1,479 @@
+    assert [event["summary"] for event in events] == event_order
+
+
+# @pytest.mark.parametrize(
```
I think it is OK to not test every corner case here. At some level, the local calendar test is exercising some `ical` functionality to make sure it actually works well, but we can trust that it mostly handles things.  I think you can remove this commented out code.

Thomas55555 replied:
https://github.com/home-assistant/core/pull/138862/commits/9440926b267988b14536d13c578fe2481ba8a419

allenporter commented on tests/components/remote_calendar/conftest.py:
```diff
@@ -0,0 +1,166 @@
+        yield client
+
+
+class Client:
```
My impression is the client fixture is only needed for mutations which are not supported by this component so you can drop everything here and below I think.

Thomas55555 replied:
removed here: https://github.com/home-assistant/core/pull/138862/commits/f7d4ce13d891e65301969223977cb1c5c5ecbea0

allenporter commented on homeassistant/components/remote_calendar/quality_scale.yaml:
```diff
@@ -0,0 +1,82 @@
+    comment: Diagnostics not implemented, yet.
+  discovery-update-info:
+    status: todo
+  discovery:
```
Is there a discovery protocol for ics urls? I would assume these would be exempt since they don't really exist.

Thomas55555 replied:
No, I changed it here: https://github.com/home-assistant/core/pull/138862/commits/21cf5e83743d0539cbf27d332a74cecdeb522a0e

allenporter commented on homeassistant/components/remote_calendar/quality_scale.yaml:
```diff
@@ -0,0 +1,82 @@
+  # Platinum
+  async-dependency: todo
+  inject-websession: done
+  strict-typing: todo
```
You might be close on this btw.

allenporter reviewed (changes requested):
Looking very good overall. I started reviewing the tests, but not sure if they were ready yet -- You can "Mark as ready" when you are ready for another round of review.

allenporter commented on homeassistant/components/remote_calendar/quality_scale.yaml:
```diff
@@ -0,0 +1,82 @@
+  parallel-updates: done
+  test-coverage: done
+  integration-owner: done
+  docs-installation-parameters: done
```
I would consider the URL an installation parameter that is not documented.  Essentially the idea is to give the user some idea of the information they are going to need when setting this up.

Thomas55555 replied:
https://github.com/home-assistant/home-assistant.io/pull/37568/commits/d77e060711a9fb05f9f2f93cb31023e376383eb4

Thomas55555 committed:
use respx

Thomas55555 committed:
just use config_entry argument that already is defined

Thomas55555 committed:
Also assert on the config entry result title and data

Thomas55555 committed:
improve config_flow

Thomas55555 committed:
update quality scale

Thomas55555 commented on tests/components/remote_calendar/test_init.py:
```diff
@@ -0,0 +1,54 @@
+    """Test update failed."""
+    mock_httpx_client.get.side_effect = exception
+    await setup_integration(hass, config_entry)
+    config_entry = hass.config_entries.async_entries(DOMAIN)[0]
```
https://github.com/home-assistant/core/pull/138862/commits/8f2e7fc00db8a86e35a4bf1641cb16bcd989d260

Thomas55555 commented on tests/components/remote_calendar/test_config_flow.py:
```diff
@@ -0,0 +1,146 @@
+            CONF_URL: CALENDER_URL,
+        },
+    )
+    assert result2["type"] is FlowResultType.CREATE_ENTRY
```
https://github.com/home-assistant/core/pull/138862/commits/f3df47ad35bf5873a9d264b5b8640e5d1296c99a

Thomas55555 commented on tests/components/remote_calendar/test_config_flow.py:
```diff
@@ -0,0 +1,146 @@
+            CONF_URL: "invalid",
+        },
+    )
+    assert result2["type"] is FlowResultType.FORM
```
https://github.com/home-assistant/core/pull/138862/commits/9440926b267988b14536d13c578fe2481ba8a419

Thomas55555 commented on homeassistant/components/remote_calendar/quality_scale.yaml:
```diff
@@ -0,0 +1,82 @@
+    status: todo
+  docs-data-update: todo
+  docs-examples: todo
+  docs-known-limitations: done
```
I changed it to todo: https://github.com/home-assistant/core/pull/138862/commits/21cf5e83743d0539cbf27d332a74cecdeb522a0e

Thomas55555 commented on homeassistant/components/remote_calendar/quality_scale.yaml:
```diff
@@ -0,0 +1,82 @@
+  docs-known-limitations: done
+  docs-supported-devices: done
+  docs-supported-functions: done
+  docs-troubleshooting: done
```
I changed it to todo: https://github.com/home-assistant/core/pull/138862/commits/21cf5e83743d0539cbf27d332a74cecdeb522a0e

Thomas55555 commented:
I changed the tests to use `respx`. 
I also tried to address your review comments (not sure, which intents, for the ics_content you prefere)
Updated the quality scale and set it to silver.
I also removed the `IcsCalendarStream.calendar_from_ics` from the config flow as it might be irritating for the use. I use the invalid ics info from local calendar there as well now.

allenporter commented on homeassistant/components/remote_calendar/strings.json:
```diff
@@ -18,19 +18,19 @@
-      "not_authorized": "You are not authorized to access this URL",
+      "unsupported_protocol": "Request URL is missing an 'http://' or 'https://' protocol.",
+      "url_not_reachable": "URL is not reachable.",
+      "cannot_connect": "You cannot connect to the URL.",
```
How about:
[%key:common::config_flow::error::cannot_connect%]

allenporter commented on homeassistant/components/remote_calendar/strings.json:
```diff
@@ -18,19 +18,19 @@
-      "url_not_reachable": "URL is not reachable",
-      "not_authorized": "You are not authorized to access this URL",
+      "unsupported_protocol": "Request URL is missing an 'http://' or 'https://' protocol.",
+      "url_not_reachable": "URL is not reachable.",
```
From a user perspective, i think this distinction is not needed and can use `cannot_connect`

allenporter commented on tests/components/remote_calendar/test_config_flow.py:
```diff
@@ -0,0 +1,208 @@
+    ("side_effect", "errors"),
+    [
+        (ConnectError("Connection failed"), "url_not_reachable"),
+        (UnsupportedProtocol("Unsupported protocol"), "unsupported_protocol"),
+        (ValueError("Invalid response"), "unknown_url_type"),
```
I think you should test these last two errors with actual input that makes them raised. For example, i'm not sure if either of these would be `InvalidURL`  instead.

Thomas55555 replied:
I tried this, and it works,:
```async def test_unsupported_protocol(hass: HomeAssistant, ics_content: str) -> None:
    """Test that an unsupported protocol results in a form error."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": SOURCE_USER}
    )
    assert result["type"] is FlowResultType.FORM

    result2 = await hass.config_entries.flow.async_configure(
        result["flow_id"],
        {
            CONF_CALENDAR_NAME: CALENDAR_NAME,
            CONF_URL: "unsupported://protocol.com",
        },
    )

    assert result2["type"] is FlowResultType.FORM
    assert result2["errors"] == {"base": "unsupported_protocol"}
```
But after that I'm not able to finish the flow, because of mocking httpx. It's a little bit difficult to explain, I asked a AI for that, with that result:
"Without using a side_effect, respx will intercept the unsupported protocol request and provide a default response rather than letting httpx throw an UnsupportedProtocol exception. As a result, the error branch in your flow isn't triggered, and the flow continues to create an entry (CREATE_ENTRY) instead of returning the expected form error. This means that without simulating the exception (via side_effect), you cannot accurately test the unsupported protocol error while still testing a successful CREATE_ENTRY later in the same flow."
Hope this helps, so we can either test it, without success in the end, or raise the side_effect. If I understand the HomeAssistant philosophy right, we wouldn't do these tests for a library like pyrainbird for example, and instead rely on, that the library is raising the right exceptions? Maybe we should also do this here?

allenporter commented on homeassistant/components/remote_calendar/config_flow.py:
```diff
@@ -0,0 +1,79 @@
+        try:
+            res = await client.get(user_input[CONF_URL], follow_redirects=True)
+            res.raise_for_status()
+        except UnsupportedProtocol as err:
```
It seems like this set of errors from the httpx exception hierarchy feel a bit arbitrary, and its not catching all of them.
Having `HTTPError` as a catchall seems useful and `TimeoutException` seem common.
I don't see that `ValueError` is actually raised. 

I think I would drop `ConnectError` and `ValueError` and makde `HTTPStatusError and `HTTPError` raise `cannot_connect`

Same for the `coordinator.py`

Thomas55555 replied:
We can even drop HTTPStatusError, as it is a subclass of HTTPError?

Thomas55555 replied:
But we should keep `InvalidURL` as it is a different Class

allenporter commented on tests/components/remote_calendar/test_config_flow.py:
```diff
@@ -34,54 +37,112 @@ async def test_form_import_ics(
+            CONF_URL: CALENDER_URL,
+        },
+    )
+    assert result3["type"] is FlowResultType.CREATE_ENTRY
```
Please test the config entry fields

allenporter commented on tests/components/remote_calendar/test_config_flow.py:
```diff
@@ -34,54 +37,112 @@ async def test_form_import_ics(
+        CONF_CALENDAR_NAME: CALENDAR_NAME,
+        CONF_URL: CALENDER_URL,
+    }
     assert result2["type"] is FlowResultType.CREATE_ENTRY
```
Please put this first by convention and I think easier to understand error message if the test fails

allenporter commented on tests/components/remote_calendar/test_config_flow.py:
```diff
@@ -92,6 +153,7 @@ async def test_no_valid_calendar(
     )
 
     assert result2["type"] is FlowResultType.FORM
```
Each test should end with `CREATE_ENTRY` or `ABORT` to verify the rest of the flow can be completed after an error.

allenporter commented on homeassistant/components/remote_calendar/strings.json:
```diff
@@ -18,19 +18,19 @@
     },
     "unable_to_parse": {
-      "message": "Unable to parse calendar data: {err}"
+      "message": "[%key:component::local_calendar::config::error::invalid_ics_file%]"
```
Users have requested more visibility in error messages from local calendar about why things failed or are invalid for local calendar import. Did what you have before work?

Thomas55555 replied:
Currently:
<img width="524" alt="image" src="https://github.com/user-attachments/assets/22d18e0c-c6b4-449c-b146-5544afba8547" />
Before:
<img width="521" alt="image" src="https://github.com/user-attachments/assets/0275fa40-3ce0-47a0-852d-6f8a6ab17c9f" />
I think the message before isn't saying much for a user?

allenporter replied:
Yeah, i think we should fix the exception to be more helpful.

Thomas55555 committed:
address review

allenporter reviewed (approved):
Looks good to me. I woudln't mind giving someone else a moment to double check, but otherwise looks good to go! Thank you for this fantastic contribution @Thomas55555 !

Thomas55555 commented:
Thanks for the cooperative review @allenporter 
Someone else looking over it sounds good.

allenporter committed:
Merge branch 'dev' into remote_calendar

allenporter commented:
There were some test flakes so I updated the branch to re-run CI.  I'll merge this when CI passes.

NoRi2909 commented on homeassistant/components/remote_calendar/strings.json:
```diff
@@ -0,0 +1,33 @@
+          "url": "Calendar URL"
+        },
+        "data_description": {
+          "calendar_name": "The name of the calendar shown in th UI.",
```
Typo here: "th" -> "the"

NoRi2909 commented on homeassistant/components/remote_calendar/strings.json:
```diff
@@ -0,0 +1,33 @@
+      "user": {
+        "description": "Please choose a name for the calendar to be imported",
+        "data": {
+          "calendar_name": "Calendar Name",
```
This should be sentence-cased -> "Calendar name"

allenporter commented:
@Thomas55555 can you address these in a follow up PR?

MartinHjelmare commented on homeassistant/components/remote_calendar/coordinator.py:
```diff
@@ -0,0 +1,67 @@
+                IcsCalendarStream.calendar_from_ics, res.text
+            )
+        except CalendarParseError as err:
+            raise UpdateFailed(
```
Test coverage for this case seem to be missing.

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
